py/visdom/server.py

```diff
@@ -928,15 +931,50 @@ def initialize(self, app):
         self.state = app.state
         self.port = app.port
         self.env_path = app.env_path
+        self.with_login = app.with_login
+        self.user_login = app.user_login
 
     def get(self, args, **kwargs):
         items = gather_envs(self.state, env_path=self.env_path)
-        self.render(
-            'index.html',
-            user=getpass.getuser(),
-            items=items,
-            active_item=''
-        )
+        if self.with_login == False:
+           self.render(
+                'index.html',
+                user=getpass.getuser(),
+                items=items,
+                active_item=''
+            ) 
+
+        if self.with_login and self.user_login['login']:
+            self.render(
+                'index.html',
+                user=getpass.getuser(),
+                items=items,
+                active_item=''
+            )
+        elif self.with_login:
+            self.render(
+                'login.html',
+                user=getpass.getuser(),
+                items=items,
+                active_item=''
+            )
+
+    def post(self, arg, **kwargs):
+        username = self.get_argument("username") 
+        password = hashlib.sha256(self.get_argument("password").encode("utf-8")).hexdigest()
+
+        if (username == self.user_login["username"]) and (password == self.user_login["password"]):
+            self.user_login["login"] = True
```

Comment:
Again, given http://www.tornadoweb.org/en/stable/guide/security.html#user-authentication, we should instead set a cookie here:

```
self.set_secure_cookie("user", self.get_argument("name"))
```

Please remove this boolean flag.

You'll also need to add

```
def get_current_user(self):
        return self.get_secure_cookie("user")
```

here: https://github.com/facebookresearch/visdom/pull/330/files#diff-ddbe57b92511a8307bcf5999eb519e6cR299